wasm.cc

```diff
@@ -0,0 +1,338 @@
+        bytes_opt operator()(const long_type_impl&) {
+            if (!result->IsInt32() && !result->IsUint32()) {
+                throw wasm::exception("Expected integer value");
+            }
+            return long_type->decompose(int64_t(v8::Integer::Cast(*result)->Value()));
+        }
+        bytes_opt operator()(const map_type_impl& m) {
+            throw wasm::exception("map type is not yet implemented in v8");
+        }
+        bytes_opt operator()(const reversed_type_impl& r) {
+            throw wasm::exception("reversed type is not yet implemented in v8");
+        }
+        bytes_opt operator()(const set_type_impl& s) {
+            throw wasm::exception("set type is not yet implemented in v8");
+        }
+        bytes_opt operator()(const short_type_impl&) {
+            throw wasm::exception("short type is not yet implemented in v8");
+        }
+        bytes_opt operator()(const simple_date_type_impl&) {
+            throw wasm::exception("simple date type is not yet implemented in v8");
+        }
+        bytes_opt operator()(const time_type_impl&) {
+            throw wasm::exception("time type is not yet implemented in v8");
+        }
+        bytes_opt operator()(const timeuuid_type_impl&) {
+            throw wasm::exception("timeuuid type is not yet implemented in v8");
+        }
+        bytes_opt operator()(const tuple_type_impl& t) {
+            throw wasm::exception("tuple type is not yet implemented in v8");
+        }
+        bytes_opt operator()(const user_type_impl& u) {
+            throw wasm::exception("user types are not yet implemented in v8");
+        }
+        bytes_opt operator()(const uuid_type_impl&) {
+            throw wasm::exception("uuid type is not yet implemented in v8");
+        }
+        bytes_opt operator()(const varint_type_impl&) {
+            throw wasm::exception("varint type is not yet implemented in v8");
+        }
+    };
+    return visit(*return_type, visitor{isolate, result});
+}
+
+seastar::future<bytes_opt> run_script(context& ctx, const std::vector<data_value>& values, data_type return_type) {
+    v8::Isolate* isolate = ctx.isolate();
+    v8::Locker locker(isolate);
+    v8::Isolate::Scope isolate_scope(isolate);
+    v8::HandleScope handle_scope(isolate);
+    v8::TryCatch try_catch(isolate);
+    v8::Local<v8::Context> local_ctx = v8::Local<v8::Context>::New(isolate, ctx.ctx);
+    v8::Context::Scope context_scope(local_ctx);
+
+    wasm_logger.debug("Running function {}", ctx.function_name);
+
+    std::vector<v8::Handle<v8::Value>> argv;
+    for (const data_value& dv : values) {
+        argv.push_back(to_v8(isolate, dv));
+    }
+
+    v8::Local<v8::Function> function = v8::Local<v8::Function>::New(isolate, ctx.function);
+    v8::Local<v8::Value> result;
+    if (!function->Call(local_ctx, local_ctx->Global(), argv.size(), argv.data()).ToLocal(&result)) {
+        throw make_wasm_exception("Running wasm function failed", isolate, try_catch);
+    }
```

Comment:
I see you thoughtfully return a future, but we need to make the run preemptible somehow. Lua supports this by counting bytecode instructions and calling a callback on overflow.

Reply:
I've been thinking about that, and I have a follow-up question - did this Lua trick prove useful in tests, or was it done as a matter of good practice? I'm asking because when it comes to UDF, simple transformations performed on a single row shouldn't usually be large enough to produce stalls - I'd imagine that it's usually refactoring a string or performing some simple math operations on the input, not actually computing things with an O(n^3) algorithm. The reason I ask is that Lua's trick also forced the call to be wrapped in a Seastar thread, which hurts its scalability - which can be alleviated a little e.g. by using a smaller stack, but it's still overhead. Perhaps the decision whether to mark the transformation preemptible could be left for users' consideration? I.e. by default the function call is not preemptible, which means it's either fast enough or the user doesn't care about occasional latency spikes (which would be common for analytical workloads, and I suspect UDF's are more useful for analytics than online queries), and the user can explicitly mark the function as heavy, which will make Scylla actually spawn a thread for it and preempt when needed.

Also, perhaps Seastar's cooperational model can be pushed down to UDF's, at least in case of javascript - it's probably possible to create a `scylla_maybe_yield()` function wrapper inside the v8 context (or any other engine), and calling this wrapper from inside a UDF would simply jump back to host and call `thread::maybe_yield()`. That would save us from counting bytecode and all. Finally, I suspect that these javascript engines won't become friends with our stall detector anyway, given that they often try just-in-time compilation and other tricks. But perhaps with enough configuration they can become more predictable.